py-rse/errors.Rmd

```diff
@@ -155,87 +154,134 @@ except ArithmeticError as error:
+We could achieve this by wrapping the open statement in a `try/except` clause:
 
 ```python
-if (val is not None) and (min_val < val < max_val) and (val % 2 == 0) and (val >= prev_val):
-    # OK branch
-else:
-    raise ValueError('Bad value: {}'.format(val))
+for file_name in args.infiles:
+    try:
+        with open(file_name, 'r') as reader:
+            update_counts(reader, word_counts)
+    except ValueError as e:
+        print(f'{file_name} is not a CSV-file.')
+        print(f'ValueError: {e}')
+```
+
+```shell
+$ python bin/collate.py Makefile
 ```
 
-Knowing what value caused the error is helpful,
-but the message in the exception doesn't tell us why it was bad.
-Was it not between `min_val` and `max_val`?
-Was it less than `prev_val`?
-Whoever has to debug the problem will want to know,
-but breaking the test up into cases makes the code harder to read:
+```text
+Makefile is not a CSV-file.
+ValueError: not enough values to unpack (expected 2, got 1)
+```
+
+This is definitely more informative than before!
+However,
+an issue with this approach is that all `ValueErrors`
+that are raised when trying to open a file
+will result in this error message,
+including those potentially raised when actually using a CSV-file as input.
+
+A more precise approach in this case would be to throw an exception specifically
+if a file other than a CSV-file is passed as input:
 
 ```python
-if val is None:
-    raise ValueError('Value {} is None'.format(val))
-elif (val <= min_val) or (max_val <= val):
-    raise ValueError('Value {} is out of range ({}, {})'.format(val, min_val, max_val))
-elif val % 2 != 0:
-    raise ValueError('Value {} is odd'.format(val))
-elif val < prev_val:
-    raise ValueError('Value {} is not less than previous value {}'.format(val, prev_val))
-else:
-    # OK branch
+for file_name in args.infiles:
+    if file_name[-4:] != '.csv':
+        raise OSError(f'{file_name} is not a CSV-file.')
+    with open(file_name, 'r') as reader:
+        update_counts(reader, word_counts)
 ```
 
+```shell
+$ python bin/collate.py Makefile                                                                          1 ↵
+```
+
+```text
+Traceback (most recent call last):
+  File "bin/collate.py", line 29, in <module>
+    main(args)
+  File "bin/collate.py", line 18, in main
+    raise OSError(f'{file_name} is not a valid CSV-file of word counts.')
+OSError: Makefile is not a CSV-file.
+```
+
+Note that this is not perfect,
+ideally we should check the content of the file
+and confirm that the format is that of the output files from `countwords.py`,
+instead of relying on the file extension.
+This could involve checking that there are two columns separated by a comma,
+that the first column contains strings,
+and that the second is numerical.
+But for the purpose of demonstration,
+our sample code above works fine.
+
+The `if/raise` approach is commonly referred to as "look before your leap",
```

Comment:
Who calls them this? (citation please)

Reply:
I just left this in from the previous version. @DamienIrving, do you know where this is from? I'm fine with removing or citing.

Reply:
No, I don't know where that came from.